relayer/crates/starknet-chain-components/src/types/cosmos/client_state.rs

```diff
@@ -101,3 +109,244 @@ impl From<CometClientState> for Any {
+            empty_child,
+            hash: hash as i32,
+        })
+    }
+}
+
+pub struct EncodeProofSpec;
+
+#[cgp_provider(MutEncoderComponent)]
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, ProofSpec> for EncodeProofSpec
+where
+    Encoding: CanEncodeMut<Strategy, Product![LeafOp, InnerSpec, u32, u32, bool]>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        proof_specs: &ProofSpec,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        let ProofSpec {
+            leaf_spec,
+            inner_spec,
+            max_depth,
+            min_depth,
+            prehash_key_before_comparison,
+        } = serde_json::to_value(proof_specs)
+            .and_then(serde_json::from_value)
+            .unwrap();
+        //.map_err(|_| Encoding::raise_error("invalid connection end"))?;
+
+        let leaf_spec = leaf_spec.unwrap();
+        let inner_spec = inner_spec.unwrap();
+        let max_depth = max_depth as u32;
+        let min_depth = min_depth as u32;
+
+        encoding.encode_mut(
+            &product![
+                leaf_spec,
+                inner_spec,
+                max_depth,
+                min_depth,
+                prehash_key_before_comparison,
+            ],
+            buffer,
+        )?;
+        Ok(())
+    }
+}
+
+#[cgp_provider(MutDecoderComponent)]
+impl<Encoding, Strategy> MutDecoder<Encoding, Strategy, ProofSpec> for EncodeProofSpec
+where
+    Encoding: CanDecodeMut<Strategy, Product![LeafOp, InnerSpec, u32, u32, bool]>
+        + CanRaiseAsyncError<&'static str>,
+{
+    fn decode_mut<'a>(
+        encoding: &Encoding,
+        buffer: &mut Encoding::DecodeBuffer<'a>,
+    ) -> Result<ProofSpec, Encoding::Error> {
+        let product![
+            leaf_spec,
+            inner_spec,
+            max_depth,
+            min_depth,
+            prehash_key_before_comparison
+        ] = encoding.decode_mut(buffer)?;
+        Ok(ProofSpec {
+            leaf_spec: Some(leaf_spec),
+            inner_spec: Some(inner_spec),
+            max_depth: max_depth as i32,
+            min_depth: min_depth as i32,
+            prehash_key_before_comparison,
+        })
+    }
+}
+pub struct EncodeProofSpecs;
+
+#[cgp_provider(MutEncoderComponent)]
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, ProofSpecs> for EncodeProofSpecs
+where
+    Encoding: CanEncodeMut<Strategy, Vec<ProofSpec>>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        proof_specs: &ProofSpecs,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        // FIXME: ibc-rs type doesn't have public fields
+
+        #[derive(serde::Deserialize)]
+        struct DummyProofSpecs(pub Vec<ProofSpec>);
+
+        let DummyProofSpecs(specs) = serde_json::to_value(proof_specs)
+            .and_then(serde_json::from_value)
+            .unwrap();
+        //.map_err(|_| Encoding::raise_error("invalid connection end"))?;
```

Comment:
this too.

Reply:
Done in 421fe1fb11ff66b6b39b9353b73f52ee40da5859